Add StandardRange::size method and use to get node degree.

arthurp opened the pull request:
Based on idea discussed here: https://github.com/KatanaGraph/katana/pull/35#issuecomment-761578291

Also fixes a stray bug.

After: #31 (simply to avoid a complex rebase)

ddn0 commented:
cc @insertinterestingnamehere

insertinterestingnamehere commented:
Missed this until just now. I'll look at it tomorrow. Thanks for the heads up!

insertinterestingnamehere commented:
Looking through this now. The reason we can't have llvm as a dependency of libgalois itself is really because shared object dependencies are transitive so that'd be the same as adding libllvm as a hard dependency of our libraries. It has relatively little to do with how our build is set up.

insertinterestingnamehere commented on CMakeLists.txt:
```diff
@@ -330,6 +330,9 @@ endif()
 find_package(Boost 1.58.0 REQUIRED COMPONENTS serialization iostreams)
 include_directories(${Boost_INCLUDE_DIRS})
 
+find_package(LLVM REQUIRED CONFIG)
```
We should still allow building the core libraris (and not the apps) without LLVM, so, for example `make libgalois_shmem` should work without LLVM present. We can tackle that later if needed though.

ddn0 replied:
Yeah, the only use of llvm in libgalois is in experimental code.

If we get a handle on https://github.com/IntelligentSoftwareSystems/Galois/issues/94 we should also remove the libgalois->libllvm dependency too.

ddn0 commented:
Closed in favor of https://github.com/IntelligentSoftwareSystems/Galois/pull/104

arthurp committed:
Add size (and a few other container methods) to StandardRange.

arthurp committed:
Fix bug in SourcePicker. Add regression test.

This fixes the same bug as
https://github.com/IntelligentSoftwareSystems/Galois/commit/088fe7576402424cbc9964b99c5bc09222421b3e

arthurp committed:
Update code to use new StandardRange::size() method.

roshandathathri commented on libgalois/src/analytics/Utils.cpp:
```diff
@@ -19,15 +19,15 @@
-  } while (
-      std::distance(graph.edges(source).begin(), graph.edges(source).end()));
+  } while (graph.edges(source).empty());
+  KATANA_LOG_ASSERT(graph.edges(source));
```
Confused by this. `edges()` returns a range, so what is this asserting?

Also, I think `assert()` (debug-only) is better here than `KATANA_LOG_ASSERT` (which is also invoked in release builds).

arthurp replied:
C++ ranges are "true-ish" if they are not empty. So this is equivalent to `graph.edges(source).size() > 0`.

Yea. Debug only would be good. But I'd like to have a KATANA macro for it. The reason is that the aborting KATANA macros will eventually actually throw an exception to abort the C++ code in a way that doesn't crash the Python interpreter. Python extensions are not supposed to abort. I'll post about this on code-related.

roshandathathri replied:
Thanks

roshandathathri reviewed (approved):
Yay. LGTM

arthurp committed:
Fix Triangle Count bug.

It wasn't non-determinism. It was an untested code path which was
enabled by the fix to IsApproximateDegreeDistributionPowerLaw.

Fixes #43

arthurp commented:
@roshandathathri I had to fix another bug (the last commit). Do you want to check it?

roshandathathri reviewed (approved):
LGTM
